src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.cs

```diff
@@ -393,17 +402,104 @@ public sealed override bool AreLocalsZeroed
+                {
+                    TypeSymbol? acyclicExtendedType = GetDeclaredExtensionUnderlyingType();
+                    if (acyclicExtendedType is not null && BaseTypeAnalysis.TypeDependsOn(acyclicExtendedType, this))
+                    {
+                        acyclicExtendedType = CyclicInheritanceError(acyclicExtendedType);
+                    }
+
+                    Interlocked.CompareExchange(ref _lazyExtendedType, acyclicExtendedType, ErrorTypeSymbol.UnknownResultType);
+                }
+
+                return _lazyExtendedType;
+            }
+        }
 
-        internal sealed override TypeSymbol? ExtensionUnderlyingTypeNoUseSiteDiagnostics => null;
         internal sealed override ImmutableArray<NamedTypeSymbol> BaseExtensionsNoUseSiteDiagnostics
-            => ImmutableArray<NamedTypeSymbol>.Empty;
+        {
+            get
+            {
+                if (_lazyBaseExtensions.IsDefault)
+                {
+                    var declaredBaseExtensions = GetDeclaredBaseExtensions();
+
+                    ImmutableArray<NamedTypeSymbol> result = declaredBaseExtensions
+                        .SelectAsArray(t => BaseTypeAnalysis.TypeDependsOn(t, on: this) ? CyclicInheritanceError(t) : t);
+
+                    ImmutableInterlocked.InterlockedCompareExchange(ref _lazyBaseExtensions, result, default);
+                }
+
+                return _lazyBaseExtensions;
+            }
+        }
 
         internal sealed override TypeSymbol? GetDeclaredExtensionUnderlyingType()
-            => throw new System.Exception("PROTOTYPE"); // PROTOTYPE
+        {
+            if (TypeKind != TypeKind.Extension)
+                return null;
+
+            if (ReferenceEquals(_lazyDeclaredExtendedType, ErrorTypeSymbol.UnknownResultType))
+            {
+                var extendedType = _underlyingType.GetDeclaredExtensionUnderlyingType();
+                var declaredExtendedType = extendedType is null ? null : this.RetargetingTranslator.Retarget(extendedType, RetargetOptions.RetargetPrimitiveTypesByName);
+                if (declaredExtendedType is not null)
+                {
+                    if (SourceExtensionTypeSymbol.AreStaticIncompatible(extendedType: declaredExtendedType, extensionType: this)
+                        || SourceExtensionTypeSymbol.IsRestrictedExtensionUnderlyingType(declaredExtendedType))
+                    {
+                        declaredExtendedType = MakeErrorType(declaredExtendedType);
+                    }
+                }
+
+                Interlocked.CompareExchange(ref _lazyDeclaredExtendedType, declaredExtendedType, ErrorTypeSymbol.UnknownResultType);
+            }
+
+            return _lazyDeclaredExtendedType;
+        }
+
+        private static NamedTypeSymbol MakeErrorType(TypeSymbol type)
+        {
+            var info = new CSDiagnosticInfo(ErrorCode.ERR_ErrorInReferencedAssembly, type.ContainingAssembly?.Identity.GetDisplayName() ?? string.Empty);
```

Comment:
>ERR_ErrorInReferencedAssembly [](http://example.com/codeflow?start=54&length=29)

The error feels too general, and quite frankly misleading. There is no error in the assembly, but the type doesn't fit the requirements given the totality of references. Perhaps ERR_MalformedExtensionInMetadata is a better fit.  Or something like: "Extension type declaration is malformed".
Fine to follow up later.

Reply:
Note the error refers to the assembly which is not the one declaring the extension type being retargeted here.